src/scalarstats.jl

```diff
@@ -577,8 +550,12 @@ function mad!(x::AbstractArray;
               normalize::Union{Bool,Nothing}=true,
               constant=nothing)
     isempty(x) && throw(ArgumentError("mad is not defined for empty arrays"))
-    x .= abs.(x .- center)
-    m = median!(x)
+    c = center === nothing ? median!(x) : center
+    T = promote_type(typeof(c), eltype(x))
+    U = eltype(x)
+    x2 = U == T ? x : isconcretetype(U) && isconcretetype(T) && sizeof(U) == sizeof(T) ? reinterpret(T, x) : similar(x, T)
```

Comment:
Let me just understand this: If `U` is an Int and `T` is `Float64`, this would allow to overwrite the `Int` array by reinterpreting it filling it with intermediate floats, but `x` interpreted as Ints is garbage afterwords. We should warn that `x` cannot be used afterwards in `mad!`

Reply:
The docstring begins with "Compute the median absolute deviation (MAD) of array x around center (by default, around the median), overwriting x in the process." so this warning does exist.

`x` can be reused, it has simply been overwritten with gibberish.

Previously, `mad!` began with `x .= abs.(x .- center)` and then partially sorted `x`, which makes the contents of `x` fairly unusable, though not quite as bizarre as `x` can end up now.

We can make the warning more visible if you'd like, but it's also contained in the function name itself. `mad!` clearly mutates its input, the warning is just that it probably mutates it in a way you aren't expecting.

Reply:
Ah, I am happy with "overwriting x in the process"